client/state/state_database_badger.go

```diff
@@ -0,0 +1,258 @@
+package state
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/dgraph-io/badger"
+	badgeropt "github.com/dgraph-io/badger/options"
+	trstate "github.com/hashicorp/nomad/client/allocrunnerv2/taskrunner/state"
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+/*
+The client has a Badger backed state store. The schema as of 0.6 looks as follows:
+
+allocations/ (bucket)
+|--> <alloc-id>/ (bucket)
+    |--> alloc -> *structs.Allocation
+    |--> alloc_runner persisted objects (k/v)
+	|--> <task-name>/ (bucket)
+        |--> task_runner persisted objects (k/v)
+*/
+
+// BadgerStateDB persists and restores Nomad client state in a Badger KV store. All
+// methods are safe for concurrent access.
+type BadgerStateDB struct {
+	db *badger.DB
+}
+
+// NewBadgerStateDB creates or opens an existing boltdb state file or returns an
+// error.
+func NewBadgerStateDB(stateDir string) (StateDB, error) {
+	opts := badger.DefaultOptions
+	opts.Dir = stateDir
+	opts.ValueDir = stateDir
+	opts.ValueLogLoadingMode = badgeropt.FileIO //no need for (default) mmap here
```

Comment:
I would leave the default as the state files shouldn't be too large, and letting the OS manage paging to/from disk seems optimal.